packages/tpetra/core/src/Tpetra_replaceDiagonalCrsMatrix_def.hpp

```diff
@@ -0,0 +1,150 @@
+// @HEADER
+
+#ifndef TPETRA_REPLACEDIAGONALCRSMATRIX_DEF_HPP
+#define TPETRA_REPLACEDIAGONALCRSMATRIX_DEF_HPP
+
+/// \file Tpetra_replaceDiagonalCrsMatrix_def.hpp
+/// \brief Definition of Tpetra::repalceDiagonalCrsMatrix
+
+#include "Tpetra_Details_Behavior.hpp"
+#include "Tpetra_CrsMatrix.hpp"
+#include "Tpetra_Vector.hpp"
+
+namespace Tpetra {
+
+template<class SC, class LO, class GO, class NT>
+LO
+replaceDiagonalCrsMatrix (::Tpetra::CrsMatrix<SC, LO, GO, NT>& matrix,
+    const ::Tpetra::Vector<SC, LO, GO, NT>& newDiag) {
+
+  using map_type = ::Tpetra::Map<LO, GO, NT>;
+  using crs_matrix_type = ::Tpetra::CrsMatrix<SC, LO, GO, NT>;
+  // using vec_type = ::Tpetra::Vector<SC, LO, GO, NT>;
+
+  const LO zeroLO = Teuchos::OrdinalTraits<LO>::zero();
+  const LO oneLO = Teuchos::OrdinalTraits<LO>::one();
+
+  // Count number of successfully replaced diagonal entries
+  LO numReplacedDiagEntries = 0;
+
+  // Extract some useful data
+  auto rowMapPtr = matrix.getRowMap();
+  if (rowMapPtr.is_null() || rowMapPtr->getComm().is_null()) {
+    // Processes on which the row Map or its communicator is null
+    // don't participate.  Users shouldn't even call this method on
+    // those processes.
+    return numReplacedDiagEntries;
+  }
+  auto colMapPtr = matrix.getColMap();
+
+  TEUCHOS_TEST_FOR_EXCEPTION
+    (colMapPtr.get () == nullptr, std::invalid_argument,
+     "replaceDiagonalCrsMatrix: "
+     "Input matrix A must have a nonnull column Map.");
+
+  const map_type& rowMap = *rowMapPtr;
+  const map_type& colMap = *colMapPtr;
+  const LO myNumRows = static_cast<LO>(matrix.getNodeNumRows());
+  const bool isFillCompleteOnInput = matrix.isFillComplete();
+
+  if (Tpetra::Details::Behavior::debug()) {
+    TEUCHOS_TEST_FOR_EXCEPTION(!rowMap.isSameAs(*newDiag.getMap()), std::runtime_error,
+        "Row map of matrix and map of input vector do not match.");
+  }
+
+  crs_matrix_type::execution_space::fence(); // for UVM's sake
+
+  if (isFillCompleteOnInput)
+    matrix.resumeFill();
+
+  Teuchos::ArrayRCP<const SC> newDiagData = newDiag.getVector(0)->getData();
+  LO numReplacedEntriesPerRow = 0;
+
+  // Loop over all local rows to replace the diagonal entry row by row
+  for (LO lclRowInd = 0; lclRowInd < myNumRows; ++lclRowInd) {
+
+    // Get row and column indices of the diagonal entry
+    const GO gblInd = rowMap.getGlobalElement(lclRowInd);
+    const LO lclColInd = colMap.getLocalElement(gblInd);
+
+    const SC vals[] = {static_cast<SC>(newDiagData[lclRowInd])};
+    const LO cols[] = {lclColInd};
+
+    // Do the actual replacement of the diagonal element
+    numReplacedEntriesPerRow = matrix.replaceLocalValues(lclRowInd, oneLO, vals, cols);
+
+    // Check for success of replacement
+    if (numReplacedEntriesPerRow == oneLO) {
+      ++numReplacedDiagEntries;
+    } else if (numReplacedEntriesPerRow == zeroLO) {
+      TEUCHOS_TEST_FOR_EXCEPTION(true, std::runtime_error,
```

Comment:
Do you really want to throw here?  What if `newDiagData[0]` is zero?  Then you've successfully replaced the value, right?  I think it would be better to return some indication of local success, and let the caller decide whether they care.